src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1199 @@
+    Lattice *ofst, bool use_final_probs, int32 frame_begin, int32 frame_end,
+    bool create_initial_state, bool create_final_state) {
+  typedef LatticeArc Arc;
+  typedef Arc::StateId StateId;
+  typedef Arc::Weight Weight;
+  typedef Arc::Label Label;
+
+  if (decoding_finalized_ && !use_final_probs)
+    KALDI_ERR << "You cannot call FinalizeDecoding() and then call "
+              << "GetRawLattice() with use_final_probs == false";
+
+  unordered_map<Token *, BaseFloat> final_costs_local;
+
+  const unordered_map<Token *, BaseFloat> &final_costs =
+      (decoding_finalized_ ? final_costs_ : final_costs_local);
+  if (!decoding_finalized_ && use_final_probs)
+    ComputeFinalCosts(&final_costs_local, NULL, NULL);
+
+  ofst->DeleteStates();
+  if (create_initial_state) ofst->AddState(); // initial-state for the chunk
+  // num-frames plus one (since frames are one-based, and we have
+  // an extra frame for the start-state).
+  KALDI_ASSERT(frame_end > 0);
+  const int32 bucket_count = num_toks_ / 2 + 3;
+  unordered_map<Token *, StateId> tok_map(bucket_count);
+  // First create all states.
+  std::vector<Token *> token_list;
+  for (int32 f = frame_begin; f <= frame_end; f++) {
+    if (active_toks_[f].toks == NULL) {
+      KALDI_WARN << "GetRawLattice: no tokens active on frame " << f
+                 << ": not producing lattice.\n";
+      return false;
+    }
+    TopSortTokens(active_toks_[f].toks, &token_list);
+    for (size_t i = 0; i < token_list.size(); i++)
+      if (token_list[i] != NULL) tok_map[token_list[i]] = ofst->AddState();
+  }
+  // The next statement sets the start state of the output FST.  Because we
+  // topologically sorted the tokens, state zero must be the start-state.
+  StateId begin_state = 0;
+  ofst->SetStart(begin_state);
+
+  KALDI_VLOG(4) << "init:" << num_toks_ / 2 + 3
+                << " buckets:" << tok_map.bucket_count()
+                << " load:" << tok_map.load_factor()
+                << " max:" << tok_map.max_load_factor();
+  // Create initial_arc for later appending with the previous chunk
+  if (create_initial_state) {
+    for (Token *tok = active_toks_[frame_begin].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      // state_label_map_ is construct during create_final_state
+      int32 id = state_label_map_.find(tok)->second; // it should exist
+      // TODO: calculate alpha but not use tot_cost
+      BaseFloat cost_offset = tok->tot_cost;
+      state_label_forward_prob_[id] = tok->tot_cost;
+      Arc arc(0, id, Weight(0, cost_offset), cur_state);
+      ofst->AddArc(begin_state, arc);
+    }
+  }
+  // Now create all arcs.
+  for (int32 f = frame_begin; f <= frame_end; f++) {
+    for (Token *tok = active_toks_[f].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      for (ForwardLinkT *l = tok->links; l != NULL; l = l->next) {
+        if (f == frame_begin && create_initial_state && l->ilabel == 0)
```

Comment:
This code is not well structured.  It' s very hard for me to verify that it's correct because it is not written in a clear way.